vertical-pod-autoscaler/docs/proposals/crash_loop_handling/README.md

```diff
@@ -0,0 +1,196 @@
+
+## Problem
+The current VPA recommender has the following problems which calls for enhancement in the recommending scheme for the VPA to try and auto-heal a component that go into crash loops due to spikes in resource usage.
+
+1. Current VPA recommender implementation considers historic usage of resources by a pod for recommending scale of memory and CPU. This method is not efficient in handling spike in resources usage. When a spike occurs, the recommender lacks information about it in historic usage to bump up the resources and scale up. This can result in frequent component outages.
+2. Stagnant CPU recommendations w.r.t increasing limits during scale up and decreasing them during scale down scenarios.
+3. OOM Kills and possible livenessProbe failures due to inefficient CPU and memory recommendation. One observation is that in a component that is in crash loop, memory recommendation is increased much more than CPU recommendations especially if OOMKills occur. But many a times, OOMKills occur due to CPU being throttled and recommending only memory scale up in such cases is not efficient. The below recommendation given from a sample where 100 periodic iterations with 10s intervals, of increasing the CPU millicores consumptions by ![resource-consumer](https://github.com/dj80hd/resource-consumer) shows clearly how the CPU recommendation has not scaled according the given requirement. It also shows how the memory is limits are increasing while CPU is remaining same for requests and limits.
+
+```
+Note: The experiment to simulate OOM Kill scenario was conducted using ![resource-consumer](https://github.com/dj80hd/resource-consumer).
+      The graphs below capture details of a container journey through constant CPU resource consumption. "consume-cpu" API from resouce-consumer was
+      used for this experiment. It was bombarded against the service endpoint of the deployment for 100 times with 5 seconds interval between every iteration to generate a sustained high load.
+```
+We can see the pod restarting once it faces resource crunch at around 14:21 - 14:45 local time in the graph timeline.
+![Container Resource Restarting](https://github.com/kallurbsk/autoscaler/blob/machine-controller-manager-provider/vertical-pod-autoscaler/docs/images/vpa-recommender/ctr_restarting_resource_crunch.png)
+
+4. For both CPU and memory the recommender suggests a percentile resource unit upgrade from the historical value of resource usage obtained from memory, VeriticalPodAutoScalerCheckpoint or Prometheus client. This does not take into account the current sudden bump in the resource which might be increasing rapidly (exponentially) over a short duration and might be outside any observed range captured in the history. VPA recommender does not suggest a better scale up and scale down for such spikes (it often recommends less than the current usage if limits are configured higher than the requests) resulting in crashes.
```

Comment:
If the problem is that the sampling rate is too low, have you considered increasing it in order to base recommendations on more granular data?

Reply:
Right. I guess the above comment describes. PLMK if you need a bit more information about this.